lib/localenv/clusterenv.go

```diff
@@ -122,11 +169,39 @@ func newClusterEnvironment(config clusterEnvironmentConfig) (*ClusterEnvironment
 		return nil, trace.Wrap(err)
 	}
 
-	objects, err := fs.New(packagesDir)
+	localObjects, err := fs.NewWithConfig(fs.Config{
+		Path: packagesDir,
+		User: config.serviceUser,
+	})
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	var objects blob.Objects
+	if config.nodeAddr == "" {
+		objects, err = fs.New(packagesDir)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		// To be able to use cluster-level package service,
+		// a node address is required. This is not available on nodes prior to upgrade
+		// with an version that did not support the necessary system metadata (node address
+		// and service user).
```

Comment:
When were these added?  Is this concern still relevant in 9.0?

Reply:
This is still applicable to 9.x.
These changes have roots in 5.5.x in the case when RPC credentials expired if the cluster was not upgraded in a year. For 5.5.x we added a specialized command to rotate them which was not ported for various reasons (not that it needs to be). As part of the problem with working with cluster packages using CLI commands and direct access to the etcd backend (as cluster package service was out of question for various reasons, the main reason is that the command needed to be able to operate even with the cluster package service inaccessible). Additionally, when working with cluster package store, it is important to make modifications to the fs using the system user instead of the effective user context (`root` most of the time) as the changes under a different user might break fs access for the k8s workloads that did not run under `root`.